src/Microsoft.ML.OnnxTransformer/OnnxSequenceType.cs

```diff
@@ -94,6 +94,12 @@ public override int GetHashCode()
         /// </summary>
         public override void Register()
         {
+            // this happens when use OnnxSequenceType attribute without specify sequence type
+            if (_elemType == null)
+            {
+                throw new Exception("Please specify sequence type when use OnnxSequenceType Attribute.");
+            }
```

Comment:
Perhaps it is better to throw ArgumentNullException here?

Reply:
Throwing just a plain `Exception` is not recommended.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types

X DO NOT throw System.Exception or System.SystemException.

However, ArgumentNullException is not correct either, since there is no argument to this method.

✓ DO throw ArgumentException or one of its subtypes if bad arguments are passed to a member.

In this case, InvalidOperationException is the most correct:

✓ DO throw an InvalidOperationException if the object is in an inappropriate state.

Reply:
IMO - it the best approach would be to remove the empty constructor and verify the `Type elemType` in the other constructor is never null.

Reply:
Yes, will do

---
In reply to: [330649635](https://github.com/dotnet/machinelearning/pull/4272#discussion_r330649635) [](ancestors = 330649635)

Reply:
Thanks, very good suggestion, InvalidOperationException is the most proper one to use here

---
In reply to: [330652629](https://github.com/dotnet/machinelearning/pull/4272#discussion_r330652629) [](ancestors = 330652629)

Reply:
Yes, I have a similar discussion with Harish about this yesterday, we agreed that remove empty constructor seems to be more ambiguous because when our customer use [OnnxSequenceType] attribute directly he/she will get error message like "method is inaccessible due to protection level". By throwing exception we can give user more specific error message. Make sense to you?

---
In reply to: [330653643](https://github.com/dotnet/machinelearning/pull/4272#discussion_r330653643) [](ancestors = 330653643)

Reply:
I would like to take Eric's suggestion and use InvalidOperationException here.

---
In reply to: [330649635](https://github.com/dotnet/machinelearning/pull/4272#discussion_r330649635) [](ancestors = 330649635)

Reply:
>  when our customer use [OnnxSequenceType] attribute directly he/she will get error message like "method is inaccessible due to protection level"

That's a compile time error - which is exactly what we want. We want to raise the error as early as possible, so the user knows they are doing something wrong.

Also, when I get a different error message than what you are displaying:

```C#
    class Program
    {
        [My]
        private int foo;
    }

    class MyAttribute : Attribute
    {
        public MyAttribute(int required)
        { }
    }
```

```
Error	CS7036	There is no argument given that corresponds to the required formal parameter 'required' of 'MyAttribute.MyAttribute(int)'	ConsoleApp80	C:\Users\eerhardt\source\repos\ConsoleApp80\Program.cs	13	Active
```

Pretty obvious what to do to fix it.

Reply:
I think Eric's suggestion is the better approach here. Please modify it as per his suggestion.

Reply:
Sounds good to me, will do

---
In reply to: [330736190](https://github.com/dotnet/machinelearning/pull/4272#discussion_r330736190) [](ancestors = 330736190)